osu.Game/Database/BackgroundDataStoreProcessor.cs

```diff
@@ -86,6 +92,69 @@ protected override void LoadComplete()
             });
         }
 
+        /// <summary>
+        /// Go through every beatmap and calculate the audio normalization values if they are missing.
+        /// </summary>
+        private void processAudioNormalization()
```

Comment:
Please follow the example set by every other method in this file...

- Storing the IDs to process ahead of time.
- Iterating through those stored IDs.
- Not nesting `.Write()` inside a `.Run()`.

Reply:
The problem with 1 and 2 is that the way that the loops work don't allow for calculating what to process ahead of time. If it were possible, I would have `totalcount` be only what needs to be processed rather than the number of total difficulties a user has. The loop populates other difficulties then skips over them in the proceeding loop. This behaviour would be hard to replicate in another loopn since it relies on the difficulties being populated during the loop. Sure It can be done but it would be hacky at best.

3rd point was done to have the beatmaps update instantly instead of being deferred until after everything is calculated. I did try to avoid nesting because common sense states that it's a bad idea, but well, the loop seems to break if I try this instead

```diff
diff --git a/osu.Game/Database/BackgroundDataStoreProcessor.cs b/osu.Game/Database/BackgroundDataStoreProcessor.cs
index fc38f3532c..d50d86b97d 100644
--- a/osu.Game/Database/BackgroundDataStoreProcessor.cs
+++ b/osu.Game/Database/BackgroundDataStoreProcessor.cs
@@ -119,37 +119,38 @@ private void processAudioNormalization()
             var notification = showProgressNotification(beatmapsCount, "Verifying loudness level for beatmaps", "all loudness levels have been verified");
             int processedCount = 0;
 
+            IQueryable<BeatmapSetInfo> beatmaps = null!;
             realmAccess.Run(r =>
             {
-                r.Refresh();
+                beatmaps = r.All<BeatmapSetInfo>().Where(b => !b.DeletePending);
+            });
 
-                foreach (BeatmapSetInfo beatmapSetInfo in r.All<BeatmapSetInfo>().Where(b => !b.DeletePending))
+            foreach (BeatmapSetInfo beatmapSetInfo in beatmaps)
+            {
+                foreach (BeatmapInfo beatmapInfo in beatmapSetInfo.Beatmaps)
                 {
-                    foreach (BeatmapInfo beatmapInfo in beatmapSetInfo.Beatmaps)
-                    {
-                        if (notification?.State == ProgressNotificationState.Cancelled)
-                            break;
-
-                        updateNotificationProgress(notification, processedCount, beatmapsCount);
-                        processedCount++;
+                    if (notification?.State == ProgressNotificationState.Cancelled)
+                        break;
 
-                        sleepIfRequired();
-                        if (beatmapInfo.AudioNormalization != null) continue;
+                    updateNotificationProgress(notification, processedCount, beatmapsCount);
+                    processedCount++;
 
-                        realmAccess.Write(realm =>
-                        {
-                            AudioNormalization audioNormalization = new AudioNormalization(beatmapInfo, beatmapSetInfo, new RealmFileStore(realmAccess, storage!));
-                            beatmapInfo.AudioNormalization = audioNormalization;
-                            realm.Add(audioNormalization.PopulateSet(beatmapInfo, beatmapSetInfo), true);
-                        });
-                        beatmapManager.GetWorkingBeatmap(beatmapInfo, true);
-                    }
+                    sleepIfRequired();
+                    if (beatmapInfo.AudioNormalization != null) continue;
 
-                    Logger.Log($"Processed audio normalization for {beatmapSetInfo.Metadata.Title}");
+                    realmAccess.Write(realm =>
+                    {
+                        AudioNormalization audioNormalization = new AudioNormalization(beatmapInfo, beatmapSetInfo, new RealmFileStore(realmAccess, storage!));
+                        beatmapInfo.AudioNormalization = audioNormalization;
+                        realm.Add(audioNormalization.PopulateSet(beatmapInfo, beatmapSetInfo), true);
+                    });
+                    beatmapManager.GetWorkingBeatmap(beatmapInfo, true);
                 }
 
-                r.Refresh();
-            });
+                Logger.Log($"Processed audio normalization for {beatmapSetInfo.Metadata.Title}");
+            }
+
+            realmAccess.Run(r => r.Refresh());
 
             Logger.Log("Finished processing audio normalization readings");
             completeNotification(notification, processedCount, beatmapsCount);
```

Reply:
As an update since the memory leak fixes:
My argument to point 1 and 2 stand. While I still want to avoid deferring the realm updates until everything is complete (this will take a LONG time for big libraries, so being able to quit but save progress would be useful), I do understand that this is more efficient. For the time being I've decided to just defer everything, but If its possible not to, I would prefer that.